dask/optimize.py

```diff
@@ -22,49 +22,60 @@ def cull(dsk, keys):
     Examples
     --------
     >>> d = {'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)}
-    >>> cull(d, 'out')  # doctest: +SKIP
+    >>> dsk, dependencies = cull(d, 'out')  # doctest: +SKIP
+    >>> dsk  # doctest: +SKIP
     {'x': 1, 'out': (add, 'x', 10)}
+    >>> dependencies  # doctest: +SKIP
+    {'x': set(), 'out': set(['x'])}
     """
     if not isinstance(keys, (list, set)):
         keys = [keys]
     nxt = set(flatten(keys))
     seen = nxt
-    while nxt:
-        cur = nxt
-        nxt = set()
-        for item in cur:
-            for dep in get_dependencies(dsk, item):
-                if dep not in seen:
-                    nxt.add(dep)
-        seen.update(nxt)
-    return dict((k, v) for k, v in dsk.items() if k in seen)
-
-
-def fuse(dsk, keys=None):
-    """ Return new dask with linear sequence of tasks fused together.
+    out = dict()
+    dependencies = dict()
+    stack = list(set(flatten(keys)))
+    while stack:
+        key = stack.pop()
+        out[key] = dsk[key]
+        deps = get_dependencies(dsk, key, as_list=True)
```

Comment:
Why `as_list=True`?  I think we can get by with `deps` and the values of `dependencies` as being sets.